File: python/heapSort.py
# heap sort implimentation using python
import math
import random


def isSort(arr):
    i = 0
    size = len(arr)
    while(i < (size-1)):
        if arr[i] > arr[i+1]:
            return False
        i += 1
    return True


def maxHeap(arr, length):
    if(length == 0):  # in case of list has only 1 value
        return

    parent = int(math.ceil(length/2)-1)

    # if(length % 2 == 0):
    #     parent = int(length / 2 - 1)
    # else:
    #     parent = int(math.floor(length/2))

    while(parent >= 0):
        tmpRoot = arr[parent]
        right = -1
        left = arr[parent*2+1]

        if (parent*2+2 <= length):
            right = arr[parent*2+2]
        # swap
        if left >= right and left > tmpRoot:
            arr[parent*2+1], arr[parent] = arr[parent], arr[parent*2+1]
        elif right > left and right > tmpRoot:
            arr[parent*2+2], arr[parent] = arr[parent], arr[parent*2+2]

        parent -= 1


def heapSort(arr):
    length = len(arr)-1

    while(length >= 0):
        maxHeap(arr, length)
        arr[0], arr[length] = arr[length], arr[0]  # value swap
        length -= 1


arr = []
n = 9000
for i in range(n):
    arr.append(int(random.random()*n))

# print(arr)
print(isSort(arr))
heapSort(arr)
print(isSort(arr))

File: Others/heapSort.cpp
// Heap sort using c++
#include <vector>
#include <iostream>
#include <cmath>
using namespace std;

void maxHeap(vector<int> &arr, int length)
{
    if (length == 0) // in case of list has only 1 value
        return;

    int parent = 0;

    if (length % 2 == 0)
        parent = length / 2 - 1;
    else
        parent = floor(length / 2);

    while (parent >= 0)
    {
        int tmpRoot = arr[parent];
        int right = -1;
        int left = arr[parent * 2 + 1];

        if (parent * 2 + 2 <= length)
            right = arr[parent * 2 + 2];
        // swap
        if (left >= right && left > tmpRoot)
            swap(arr[parent * 2 + 1], arr[parent]);
        else if (right >= left && right > tmpRoot)
            swap(arr[parent * 2 + 2], arr[parent]);

        parent -= 1;
    }
}

void heapSort(vector<int> &arr)
{
    int length = arr.size() - 1;

    while (length >= 0)
    {
        maxHeap(arr, length);
        swap(arr[0], arr[length]);
        length -= 1;
    }
}

int main()
{
    vector<int> arr;
    int n = 90000;
    for (size_t i = 0; i < n; i++)
    {
        arr.push_back(random() % n);
    }

    heapSort(arr);
    // for (size_t i = 0; i < n; i++)
    // {
    //     cout << arr[i] << endl;
    // }

    return 0;
}
